File: src/db/seed.ts
/**
 * Seed script: creates a demo project with everything wired so the UI has
 * something to show: the default catalog, a feature-template task, a goal,
 * a trigger, and an automation.
 *
 *   pnpm run seed   (against a fresh data dir, or after starting the server)
 *
 * The script talks to a running control plane via its operator token.
 */
import { automations } from "./schema.js";
import { loadConfig } from "../config.js";
import { openMigratedDb } from "../db/client.js";
import { ProjectService } from "../services/projects.js";
import { TaskService } from "../services/tasks.js";
import { FileService } from "../services/files.js";
import { GoalService } from "../services/goals.js";
import { InboxService } from "../services/inbox.js";
import { SessionService } from "../services/sessions.js";
import { ActivityService } from "../services/activity.js";
import { PushService } from "../services/push.js";
import { SecretService } from "../services/secrets.js";
import { SchedulerService } from "../services/scheduler.js";
import { TriggerService } from "../services/triggers.js";
import type { Services } from "../services/registry.js";

export async function main(): Promise<void> {
  const config = loadConfig();
  const db = openMigratedDb(config);
  const services: Services = {
    db,
    config,
    projects: new ProjectService(db),
    tasks: new TaskService(db),
    files: new FileService(db, config),
    goals: new GoalService(db),
    inbox: new InboxService(db),
    activity: new ActivityService(db),
    push: new PushService(db, config),
    secrets: new SecretService(db, config.secret),
    sessions: undefined as unknown as SessionService,
    scheduler: undefined as unknown as SchedulerService,
    triggers: undefined as unknown as TriggerService,
  };
  services.scheduler = new SchedulerService(db, services, undefined, 1000);
  services.sessions = new SessionService(db, config, services);
  services.triggers = new TriggerService(db, services);

  const existing = await services.projects.list();
  if (existing.some((p) => p.slug === "demo")) {
    console.log("demo project already exists — delete ./data to reseed");
    return;
  }

  const demo = await services.projects.create({ name: "demo" });
  console.log(`project: ${demo.name} (${demo.id})`);

  const agentByName = new Map(
    (await db.query.agents.findMany({ where: (t, { eq }) => eq(t.projectId, demo.id) })).map((a) => [a.name, a]),
  );
  const defaultAgent = agentByName.get("default")!;
  const linkedinAgent = agentByName.get("linkedin-content")!;

  // 1) A feature-template task (spec → plan → … → human merge).
  const tpl = (
    await db.query.taskTemplates.findMany({ where: (t, { eq }) => eq(t.projectId, demo.id) })
  ).find((t) => t.name === "compound-engineer-workflow")!;
  const chain = await services.tasks.instantiateTemplate(demo.id, tpl.id, {
    branchName: "feat/dark-mode",
    featureTitle: "Dark mode",
  });
  console.log(`template instantiated → ${chain.length} tasks (step 1: "${chain[0]!.name}", approval-gated)`);

  // 2) A simple immediate task for the default agent.
  await services.tasks.create(demo.id, {
    name: "Update onboarding docs",
    description: "Refresh the README onboarding section with the new CLI flags.",
    assigneeAgentId: defaultAgent.id,
  });
  console.log("task: Update onboarding docs (runs immediately)");

  // 3) A goal with a 2-item DoD (approve in the UI to start the loop).
  const goal = await services.goals.create({
    projectId: demo.id,
    title: "Improve error messages",
    spec: "Make runtime errors actionable.\n- Errors include the failing operation\n- Errors suggest the next step",
    spendCapUsd: 5,
  });
  console.log(`goal: "${goal.title}" (${goal.definitionOfDone.length} DoD items — approve in the UI to start)`);

  // 4) A support-inbound trigger (HMAC-signed webhook).
  const trigger = await services.triggers.create({
    projectId: demo.id,
    name: "support-inbound",
    agentId: agentByName.get("customer-support")!.id,
    jobPrompt: "Support conversation:\n{{payload}}",
  });
  console.log(`trigger: support-inbound → POST /hooks/${trigger.id}`);
  console.log(`  signature: x-agentos-signature = sha256 hmac of the raw body with secret ${trigger.webhookSecret}`);

  // 5) A weekly automation.
  await db
    .insert(automations)
    .values({
      id: crypto.randomUUID(),
      projectId: demo.id,
      name: "weekly-linkedin",
      cron: "0 9 * * 1",
      timezone: "UTC",
      agentId: linkedinAgent.id,
      taskTemplateId: null,
      taskBody: "Draft this week's LinkedIn post.",
    })
    .run();
  console.log("automation: weekly-linkedin (Mon 09:00 UTC)");

  // Kick the scheduler once so immediate tasks start immediately.
  await services.scheduler.tick();
  console.log("\nseeded — open the UI, approve the goal's DoD, and watch the loop.");
}

if (import.meta.url === `file://${process.argv[1]}`) {
  void main().catch((e) => {
    console.error(e);
    process.exit(1);
  });
}

File: agentos/db/seed.py
"""
Seed script: creates a demo project with the default agent catalog, a
feature-template task chain, a goal, a trigger, and a weekly automation.

    python -m agentos.db.seed
"""
import asyncio
import uuid


async def main() -> None:
    from agentos.config import load_config
    from agentos.db.client import create_tables, get_session, init_engine
    from agentos.db.models import Automation
    from agentos.services.registry import build_services

    config = load_config()
    init_engine(config.db_path)
    await create_tables()

    services = await build_services(config)

    existing = await services.projects.list()
    if any(p["slug"] == "demo" for p in existing):
        print("demo project already exists — delete ./data to reseed")
        return

    demo = await services.projects.create({"name": "demo"})
    print(f"project: {demo['name']} ({demo['id']})")

    agents_list = await services.projects.list_agents(demo["id"])
    agent_by_name = {a["name"]: a for a in agents_list}
    default_agent = agent_by_name.get("default")
    linkedin_agent = agent_by_name.get("linkedin-content")
    support_agent = agent_by_name.get("customer-support")

    # 1) Feature-template task chain
    templates = await services.tasks.list_templates(demo["id"])
    tpl = next((t for t in templates if t["name"] == "compound-engineer-workflow"), None)
    if tpl:
        chain = await services.tasks.instantiate_template(
            demo["id"], tpl["id"],
            {"branchName": "feat/dark-mode", "featureTitle": "Dark mode"},
        )
        print(f"template instantiated → {len(chain)} tasks")

    # 2) Simple immediate task
    if default_agent:
        await services.tasks.create(demo["id"], {
            "name": "Update onboarding docs",
            "description": "Refresh the README onboarding section with the new CLI flags.",
            "assigneeAgentId": default_agent["id"],
        })
        print("task: Update onboarding docs")

    # 3) Goal with spend cap
    goal = await services.goals.create({
        "projectId": demo["id"],
        "title": "Improve error messages",
        "spec": "Make runtime errors actionable.\n- Errors include the failing operation\n- Errors suggest next step",
        "spendCapUsd": 5,
    })
    print(f"goal: \"{goal['title']}\" ({len(goal['definitionOfDone'])} DoD items)")

    # 4) HMAC-signed webhook trigger
    if support_agent:
        trigger = await services.triggers.create({
            "projectId": demo["id"],
            "name": "support-inbound",
            "agentId": support_agent["id"],
            "jobPrompt": "Support conversation:\n{{payload}}",
        })
        print(f"trigger: support-inbound → POST /hooks/{trigger['id']}")

    # 5) Weekly automation
    if linkedin_agent:
        async with get_session() as db:
            db.add(Automation(
                id=str(uuid.uuid4()),
                project_id=demo["id"],
                name="weekly-linkedin",
                cron="0 9 * * 1",
                timezone="UTC",
                agent_id=linkedin_agent["id"],
                task_body="Draft this week's LinkedIn post.",
            ))
            await db.commit()
        print("automation: weekly-linkedin (Mon 09:00 UTC)")

    await services.scheduler.tick()
    print("\nseeded — open the UI, approve the goal DoD, and watch the loop.")


if __name__ == "__main__":
    asyncio.run(main())
